[2/N] Add flag to control which rank should perform NaN check

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134707
* #134701
* #134357
* __->__ #134345

Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[2/N] Add flag to control which rank should perform NaN check

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3097,7 +3100,9 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
-  if (enableNanCheck_) {
+  // Only check for NaN for send ops, for recv ops `tensor` can be a random
+  // placeholder
+  if (enableNanCheck_ && opType == OpType::SEND) {
```
double checking- does this check work correctly for coalesced p2p too? would this function be called when creating the individual isend ops and check their nans at that time?

kwen2501 replied:
Yeah, I think individual ops in batch P2P would call into this function too, with OpType::Send or OpType::Recv, so they should be covered.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3455,14 +3462,16 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::broadcast(
   // avoidRecordStreams_ note: collective() will stash tensors.
   bool avoidRecordStreams = avoidRecordStreams_ || (!opts.asyncOp);
 
+  const auto root = opts.rootRank + opts.rootTensor;
```
what is opts.rootTensor?

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -728,7 +728,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
       const char* profilingTitle = nullptr,
-      bool avoidRecordStreams = false);
+      bool avoidRecordStreams = false,
+      bool nanCheck = true);
```
so we default to true which makes sense, bc most collectives (allreduce, allgather, gather, a2a) will need to check all inputs.  The ones that don't are special-cased (broadcast, scatter).  i think we covered them all.  
 
Is `_broadcast_oop` covered by broadcast?

kwen2501 replied:
Thanks for checking. Added ` _broadcast_oop` now.

wconstab commented:
this one LGTM. just add tests.

kwen2501 committed:
Update on "[2/N] Add flag to control which rank should perform NaN check"


Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[2/N] Add flag to control which rank should perform NaN check"


Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[2/N] Add flag to control which rank should perform NaN check"


Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[2/N] Add flag to control which rank should perform NaN check"


Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
> this one LGTM. just add tests.

Added a test doing P2P where NaN number exists in receive buffer -- NaN checker should not bark.

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge -f "The revert is unfortunately a miss-target; see [D61938556](https://www.internalfb.com/diff/D61938556); re-landing"

kwen2501 committed:
Update on "[2/N] Add flag to control which rank should perform NaN check"


Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 committed:
Update on "[2/N] Add flag to control which rank should perform NaN check"


Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -f "previously landed, reverted base PR, re-open, CI all green, rebase, land again"
